torch/onnx/symbolic_opset14.py

```diff
@@ -117,3 +119,174 @@ def quantized_hardswish(g: jit_utils.GraphContext, x, op_scale, op_zero_point):
     output = hardswish(g, x)
 
     return symbolic_helper.quantize_helper(g, output, op_scale, op_zero_point)
+
+
+# Ported from https://github.com/microsoft/onnx-script/blob/main/onnxscript/function_libs/torch_aten/ops/core.py
+# aten_scaled_dot_product_attention
+# NOTE: Need Trilu
+@_onnx_symbolic("aten::scaled_dot_product_attention")
+@symbolic_helper.parse_args("v", "v", "v", "v", "f", "b", "f")
+@_beartype.beartype
+def scaled_dot_product_attention(
+    g: jit_utils.GraphContext,
+    query: torch._C.Value,
+    key: torch._C.Value,
+    value: torch._C.Value,
+    attn_mask: Optional[torch._C.Value] = None,
+    dropout_p: float = 0.0,
+    is_causal: bool = False,
+    scale: Optional[float] = None,
+):
+    # Use trace_only to handle optional inputs
+    assert (not is_causal) or (
+        is_causal and symbolic_helper._is_none(attn_mask)
+    ), "is_causal and attn_mask cannot be set at the same time"
+
+    if scale is None:
+        scale = _attention_scale(g, query)
+
+    if is_causal:
+        attn_mask = _causal_attention_mask(g, query, key)
+
+    # Swap the last two axes of key
+    key_shape = g.op("Shape", key)
+    end_idx = g.op(
+        "Constant", value_t=torch.tensor([_constants.INT64_MAX], dtype=torch.int64)
+    )
+    last_idx = g.op("Constant", value_t=torch.tensor([-1], dtype=torch.int64))
+    second_last_idx = g.op("Constant", value_t=torch.tensor([-2], dtype=torch.int64))
+    key_last_dim = g.op("Slice", key_shape, last_idx, end_idx)
+    key_second_last_dim = g.op("Slice", key_shape, second_last_idx, last_idx)
+    key_first_dims = g.op(
+        "Slice",
+        key_shape,
+        g.op(
+            "Constant", value_t=torch.tensor([_constants.INT64_MIN], dtype=torch.int64)
+        ),
```

Comment:
This whole block is for transpose. Would it be possible to simply find the rank of the input and swap the last axes?

Reply:
Yes, I wonder maybe Gather might be able to save this?

Reply:
Possibly? I haven't thought about using Gather carefully. A naive thing to do is just to get the rank with a symbolic helper, Then create 

```python
tranposed_axes = list(range(rank))
tranposed_axes[-1], tranposed_axes[-2] =  tranposed_axes[-2], tranposed_axes[-1]
op.Transpose(key, perms=[tranposed_axes]
```

Reply:
But the swap would be off graph? I am not sure if it's legit.

Reply:
As long as the rank doesn't change it would be ok. That is true even for dynamic axes

Reply:
> Possibly? I haven't thought about using Gather carefully. A naive thing to do is just to get the rank with a symbolic helper, Then create
> 
> ```python
> tranposed_axes = list(range(rank))
> tranposed_axes[-1], tranposed_axes[-2] =  tranposed_axes[-2], tranposed_axes[-1]
> g.op("Transpose", key, perms=[tranposed_axes])
> ```

Rethink on this. I think it won't work. Let's say we have rank from op.Shape and Op.Size, and then this onnx value wouldn't be consumed by any following ops. The inputs/attributes we feed to op.Transpose would be all generated off-graph by symbolic_helper.py, and it would all be constants in final graph.

Reply:
If that's the case, should onnx-script side make the same change?

Reply:
Semantically I think this logic should stay. Also, The rank could change?

Reply:
I debated with myself and decided that the torchlib version should be as dynamic as possible, as we are trying to match the function with the aten op. It also avoids a trace_only need. Long term we really should have a swap axis op: https://github.com/onnx/onnx/issues/4977

Reply:
> Semantically I think this logic should stay. Also, the rank could change?

That's fine too. I just worried about speed but maybe it's not much of a slow down

Reply:
Also I don't really think rank can change for the export scenarios we support. Especially with many symbolic functions looking for rank already in their logic. Are there cases where you see rank can change?

Reply:
I guess based on the idea that **rank changes across models**, promoting perms to inputs would make the code more reusable I think.

Reply:
Done